src/parse/asp/builtins.go

```diff
@@ -155,6 +159,57 @@ func buildRule(s *scope, args []pyObject) pyObject {
 	return pyString(":" + target.Label.Name)
 }
 
+// doExec fork/exec's a command and returns the output as a string.  exec
+// accepts either a string or a list of commands and arguments.
+func doExec(s *scope, args []pyObject) pyObject {
+	cmdIn := args[0]
+	wantStdout := args[1].IsTruthy()
+	wantStderr := args[2].IsTruthy()
+
+	var cmdArgs []string
+	if isType(cmdIn, "str") {
+		cmdArgs = strings.Fields(string(cmdIn.(pyString)))
+	} else if isType(cmdIn, "list") {
+		pl := cmdIn.(pyList)
+		cmdArgs = make([]string, 0, pl.Len())
+		for i := 0; i < pl.Len(); i++ {
+			cmdArgs = append(cmdArgs, pl[i].String())
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+
+	// TODO(seanc@): Memoize the output based on cmdArgs
```

Comment:
This is a good idea (both for performance and to prevent the reliance on changing commands), probably reinforces the point that we should call out the deterministic contract on the command output explicitly in the documentation above (as it in the future may be cached and reused)

Reply:
Done, though I'm not wild about the complexity this introduced.

Reply:
Thanks, I was actually just agreeing with your aspirations, did not mean for you do it it straight away :), good effort!

Reply:
I didn't like what I pushed re: promises and reworked it into something that I actually think is significantly more readable and maintainable.